### PR TITLE
feat: introduce "exit spans" and mark S3 spans as exit spans

### PR DESCRIPTION
**Pull request by trentm**
Per https://github.com/elastic/apm/blob/main/specs/agents/tracing-spans.md#exit-spans
the result is that HTTP child spans of S3 spans are no longer captured.

Closes: #2125

### Overview

There is a boolean `exitSpan` option to the `startSpan()` APIs now, which we'll eventually use in our instrumentations. For now I've just added it to the `s3` instrumentation to get a feel. When creating a new span, if its parent is an exitSpan then the span won't be created.

I manually tested that context-propagation still works, but don't have a test case for that, because it is a pain in the S3 tests. I plan to add tests for context-propagation still working in https://github.com/elastic/apm-agent-nodejs/issues/2000 (because I already have a MockES in tests that I can easily use to test the `traceparent` et al headers).

In subsequent PRs we can work through adding `, { exitSpan: true }` to the relevant instrumentations.

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Update TypeScript typings
- [x] Update documentation
- [x] Add CHANGELOG.asciidoc entry
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/main/CONTRIBUTING.md#commit-message-guidelines)
- [x] aws-sdk TAV test run

### CODE REVIEW

**Comment by trentm** [comment]
I'm experimenting a bit with adding some `span.context.http`. For example:

```
    {
        "span": {
            "name": "S3 GetObject elasticapmtest-bucket-20220128213346",
            "type": "storage",
            "id": "4efe60d5f914412c",
            "transaction_id": "766ebd597ecbc0dd",
            "parent_id": "766ebd597ecbc0dd",
            "trace_id": "32663fe436de85979ae7426674892ff4",
            "subtype": "s3",
            "action": "GetObject",
            "timestamp": 1643405628127090,
            "duration": 270.786,
            "context": {
                "http": {
                    "method": "GET",
                    "status_code": 304,
                    "url": "https://s3.us-east-2.amazonaws.com/",
                    "response": {
                        "encoded_body_size": 0,
                        "headers": {
                            "x-amz-id-2": "qN15KVu9Soe42YSRcU1SNJ8z5XByyC85fS1GDvpZS9BnJTlyHbROwebKhq3LsfuHsFhNNMZGRZU=",
                            "x-amz-request-id": "G4TW207TJAMB962W",
                            "date": "Fri, 28 Jan 2022 21:33:49 GMT",
                            "last-modified": "Fri, 28 Jan 2022 21:33:48 GMT",
                            "etag": "\"fd33e2e8ad3cb1bdd3ea8f5633fcf5c7\"",
                            "server": "AmazonS3"
                        }
                    }
                },
                "destination": {
                    "service": {
                        "name": "s3",
                        "type": "storage",
                        "resource": "elasticapmtest-bucket-20220128213346"
                    },
                    "address": "s3.us-east-2.amazonaws.com",
                    "port": 443,
                    "cloud": {
                        "region": "us-east-2"
                    }
                }
            },
            "sync": false,
            "outcome": "success",
            "sample_rate": 1
        }
    }
```

However, I am inclined to *not* include:

- method: Not that helpful.
- url: Mostly redundant with context.destination.address.
- response.headers: A lot of added size for uncertain utility. The inclusion of Amazon's request ID headers might be worth it.

**Comment by trentm** [comment]
run module tests for aws-sdk

**Comment by trentm** [comment]
The TAV test run for aws-sdk is getting out of hand again. They are a lot of the slowest TAV runs:

```
% ./dev-utils/jenkins-build-slow-steps.sh https://apm-ci.elastic.co/job/apm-agent-nodejs/job/apm-agent-nodejs-mbp/job/main/18/
...
1359145 FINISHED SUCCESS .ci/scripts/test.sh -b "release" -t "bluebird" "12"
1361483 FINISHED SUCCESS .ci/scripts/test.sh -b "release" -t "graphql" "12"
1363566 FINISHED SUCCESS .ci/scripts/test.sh -b "release" -t "aws-sdk" "8"
1386958 FINISHED SUCCESS .ci/scripts/test.sh -b "release" -t "aws-sdk" "14"
1402172 FINISHED SUCCESS .ci/scripts/test.sh -b "release" -t "pg" "12"
1405902 FINISHED SUCCESS .ci/scripts/test.sh -b "release" -t "hapi" "12"
1421825 FINISHED SUCCESS .ci/scripts/test.sh -b "release" -t "knex" "12"
1422159 FINISHED SUCCESS .ci/scripts/test.sh -b "release" -t "knex" "10"
1458659 FINISHED SUCCESS .ci/scripts/test.sh -b "release" -t "knex" "14"
1505502 FINISHED SUCCESS .ci/scripts/test.sh -b "release" -t "aws-sdk" "12"
1600148 FINISHED SUCCESS .ci/scripts/test.sh -b "release" -t "apollo-server-express" "14"
1625149 FINISHED SUCCESS .ci/scripts/test.sh -b "release" -t "apollo-server-express" "12"
1678785 FINISHED SUCCESS .ci/scripts/test.sh -b "release" -t "aws-sdk" "10"
```

Current TAV runs are testing all these release versions:

```
-- package versions matching "2.858.0 || 2.889.0 || 2.920.0 || 2.951.0 || 2.982.0 || 2.1013.0 || 2.1016.0 || >2.1016.0 <3": 2.858.0, 2.889.0, 2.920.0, 2.951.0, 2.982.0, 2.1013.0, 2.1016.0, 2.1017.0, 2.1018.0, 2.1019.0, 2.1020.0, 2.1021.0, 2.1022.0, 2.1023.0, 2.1024.0, 2.1025.0, 2.1026.0, 2.1027.0, 2.1028.0, 2.1029.0, 2.1030.0, 2.1031.0, 2.1032.0, 2.1033.0, 2.1034.0, 2.1035.0, 2.1036.0, 2.1037.0, 2.1038.0, 2.1039.0, 2.1040.0, 2.1041.0, 2.1042.0, 2.1043.0, 2.1044.0, 2.1045.0, 2.1046.0, 2.1047.0, 2.1048.0, 2.1049.0, 2.1050.0, 2.1051.0, 2.1052.0, 2.1053.0, 2.1054.0, 2.1055.0, 2.1056.0, 2.1057.0, 2.1058.0, 2.1059.0, 2.1060.0, 2.1061.0, 2.1062.0, 2.1063.0, 2.1064.0, 2.1065.0, 2.1066.0
```

I'll update the versions to test:

```
% ./dev-utils/aws-sdk-tav-versions.sh
  # Test v2.858.0, every N=41 of 209 releases, and current latest.
  versions: '2.858.0 || 2.899.0 || 2.940.0 || 2.981.0 || 2.1022.0 || 2.1063.0 || 2.1066.0 || >2.1066.0 <3'
```

**Comment by trentm** [comment]
run module tests for aws-sdk

File: lib/instrumentation/index.js [hunk 1]
```diff
@@ -83,7 +83,7 @@ Instrumentation.prototype.currTransaction = function () {
   if (!this._started) {
     return null
   }
-  return this._runCtxMgr.active().currTransaction() || null
+  return this._runCtxMgr.active().currTransaction()
```
Comment (trentm) [inline lib/instrumentation/index.js:86]: REVIEW NOTE: This was redundant. `RunContext.currTransaction()` already defaults to `null` if there isn't a transaction.

**Review by astorm — APPROVED** [review]

**Pull request merged**